### PR TITLE
Yeoman

### LINKED ISSUE #58: yeoman
do it

### PR DESCRIPTION
**Pull request by jmeas**
Resolves #58 

===

- [x] Differentiate `package.json`s
- [x] Move actual `deps` from `devDeps` in generator package.json
- [x] Ensure all files are copied over
- [x] Update variables in those files
- [x] Figure out API for not exporting a global
- [x] Add generator tests
- [x] Update Travis test for the generator
- [x] Fix Travis (//cc @thejameskyle / @megawac )
- [x] Update README
- [x] Remove global var option
- [x] Fix license escaping
- [x] Update name of initial test

### CODE REVIEW

**Comment by jmeas** [comment]
@thejameskyle, I can't seem to get travis to run. Any clue? [Here's the output from travis.](https://travis-ci.org/babel/babel-library-boilerplate/jobs/73886589#L257)

**Comment by thejameskyle** [comment]
Well the actual error is [below](https://travis-ci.org/babel/babel-library-boilerplate/jobs/73886589#L262). The problem is that yeoman is trying to execute `npm whoami` but travis does not have a user.

**Comment by thejameskyle** [comment]
Are we going to rename this repo to generator-babel-library or something?

**Comment by jmeas** [comment]
> Are we going to rename this repo to generator-babel-library or something?

Yeah, something like that. I might contact the guy who owns `generator-babel` to see if he would give me that package name.

> Well the actual error is below. The problem is that yeoman is trying to execute npm whoami but travis does not have a user.

Hmm. Were you able to get this working in your branch? Maybe I'll add a try/catch around it.

```js
function getDefaultName() {
  var name;
  try {
    name = npm.whoami();
  } catch(e) {}
  return name;
}
```

**Comment by thejameskyle** [comment]
> Maybe I'll add a try/catch around it.

:+1:

File: app/index.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      this.template('src/index.js', 'src/' + this.repo + '.js');
+      mkdirp.sync('test');
+      this.template('test/eslintrc', 'test/.eslintrc');
+      this.template('test/runner.html', 'test/runner.html');
+      mkdirp.sync('test/setup');
+      this.template('test/setup/browserify.js', 'test/setup/browserify.js');
+      this.template('test/setup/node.js', 'test/setup/node.js');
+      this.template('test/setup/setup.js', 'test/setup/setup.js');
+      mkdirp.sync('test/unit');
+      this.template('test/unit/index.js', 'test/unit/' + this.repo + '.js');
```
Comment (megawac) [inline app/index.js]: Didn't want to do a loop after a `fs.readdir()`?

File: app/templates/_package.json [hunk 2]
```diff
@@ -0,0 +1,69 @@
+{
```
Comment (megawac) [inline app/templates/_package.json]: Why not just call it `package.json`

File: app/templates/_package.json [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  "babelBoilerplateOptions": {
+    "entryFileName": "<%= repo %>",
+    "mainVarName": "<%= variable %>",
+    "exportGlobal": <%= global %>,
+    "mochaGlobals": [
+      "stub",
+      "spy",
+      "expect"
+    ]
+  }
```
Comment (megawac) [inline app/templates/_package.json]: Sticking with for now in that case?

File: app/templates/gitignore [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# Logs
```
Comment (megawac) [inline app/templates/gitignore]: **`.gitignore`**

**Comment by megawac** [comment]
Other than some of the file names looks good

File: app/templates/_package.json (hunk 2 above)
Comment (thejameskyle) [inline app/templates/_package.json]: yeoman style thing

File: app/templates/gitignore (hunk 4 above)
Comment (thejameskyle) [inline app/templates/gitignore]: another yeoman naming convention

File: app/index.js (hunk 1 above)
Comment (jmeas) [inline app/index.js]: That would work. I just copy-pasted from James' generator. I don't have a strong opinion tbqh. Do you prefer a loop?

File: app/templates/_package.json (hunk 3 above)
Comment (jmeas) [inline app/templates/_package.json]: Probably not. I just haven't updated it in the PR yet.

File: app/index.js (hunk 1 above)
Comment (megawac) [inline app/index.js]: Seems more flexible - anyway thats what I did IIRC in `generator-babel-node`

**Comment by megawac** [comment]
I was also thinking, would it make sense to add a prompt like project type: `node` (babel-node-boiler) `browser/both` (babel-library-boiler) and select scaffolds accordingly. That way we can merge `generator-babel-node` and this one

**Comment by megawac** [comment]
> > Well the actual error is below. The problem is that yeoman is trying to execute npm whoami but travis does not have a user.

> Hmm. Were you able to get this working in your fork? Maybe I'll add a try/catch around it.
```js
function getDefaultName() {
  var name;
  try {
    name = npm.whoami();
  } catch(e) {}
  return name;
}
```

[I got around this using promises](https://github.com/megawac/generator-babel-node/blob/master/app/index.js#L30)

```js
    Promise.all([gitConfig(), exec('npm whoami').catch(function(e) {
      console.error('Error getting npm user name: run `npm login`');
      console.error(e);
    })])
    .then(function(args) {
      this.config = args[0];
      this.username = trim(args[1][0]);
      this._showPrompts(done);
    }.bind(this));
```

File: app/index.js (hunk 1 above)
Comment (jmeas) [inline app/index.js]: :+1: will-change

**Comment by jmeas** [comment]
> would it make sense to add a prompt like project type

Ya I plan to do that and a few other things in a future iteration. This is just a first version tho'.

**Comment by jmeas** [comment]
> I got around this using promises

Ah, cool. I think the try-catch is a bit less ugly, so I'm going to try that first.

**Comment by jmeas** [comment]
SUCCESS.

thanks for the help @megawac.

**Comment by jmeas** [comment]
@megawac I like your simplified copying of files, but I want to stick with the Yeoman conventions for now. I may revisit later. I just want to get this first version merged 4now.

File: app/index.js (hunk 1 above)
Comment (megawac) [inline app/index.js]: Thoughts james?

Comment (jmeas) [inline app/index.js]: https://github.com/babel/generator-babel-boilerplate/pull/197#issuecomment-127468961

tl;dr, yes, but maybe in an iteration.

Comment (jmeas) [inline app/index.js]: @megawac why not just `this.directory(this.root);`?
